### ISSUE TITLE
Fuzz: assert during DISTINCT execution

### ISSUE THREAD

**Issue by devcrafter** [ClickHouse/ClickHouse#40947]
**Describe the bug**
Crash in debug build due to assert()

**How to reproduce**
Original report - https://s3.amazonaws.com/clickhouse-test-reports/40305/1b6e7b9ca21954daa58795befcb0c39de1e3df88/fuzzer_astfuzzerdebug//report.html

***Query & crash stack***:
```
Fuzzing step 14 out of 1000

SELECT DISTINCT
    ignore(if(1025, -2147483648, -2147483647), ignore(1048577)),
    JSONExtractRaw(concat('{"x":', if(rand() % 2, 'true', 'fal\0e'), '}}'), if(1025, ignore(NULL, ignore(ignore(NULL), NULL)), 1.1754943508222875e-38), 'x') AS res
FROM numbers(1000000)
ORDER BY res ASC NULLS FIRST

Timeout exceeded while receiving data from server. Waited for 10 seconds, timeout is 10 seconds.
[ip-172-31-92-10] 2022.08.19 16:09:08.979659 [ 546 ] <Fatal> BaseDaemon: ########################################
[ip-172-31-92-10] 2022.08.19 16:09:08.980413 [ 546 ] <Fatal> BaseDaemon: (version 22.8.1.1, build id: A5320CF4887A3BCF) (from thread 545) (query_id: cca5c983-e1ee-4c6a-84b0-d6f08bc096d8) (query: SELECT DISTINCT ignore(if(1025, -2147483648, -2147483647), ignore(1048577)), JSONExtractRaw(concat('{"x":', if(rand() % 2, 'true', 'fal\0e'), '}}'), if(1025, ignore(NULL, ignore(ignore(NULL), NULL)), 1.1754943508222875e-38), 'x') AS res FROM numbers(1000000) ORDER BY res ASC NULLS FIRST) Received signal Aborted (6)
[ip-172-31-92-10] 2022.08.19 16:09:08.980758 [ 546 ] <Fatal> BaseDaemon: 
[ip-172-31-92-10] 2022.08.19 16:09:08.995766 [ 546 ] <Fatal> BaseDaemon: Stack trace: 0x7f182af3700b 0x7f182af16859 0x7f182af16729 0x7f182af27fd6 0x28ec873e 0x28ec474d 0x22d8e8e2 0x28aa4161 0x28ae55a3 0x28ae5263 0x28ac4b01 0x28ac4e37 0x28ac3dd8 0x28ac34f8 0x28aec74a 0x28aec620 0x28aec5b5 0x28aec561 0x28aec472 0x28aec37a 0x28aec2b5 0x28aec27d 0x28aec255 0x28aec220 0x17295c66 0x17292195 0x1734fd1d 0x173565e4 0x17356575 0x173564a5 0x17355e02 0x7f182b0ee609 0x7f182b013133
[ip-172-31-92-10] 2022.08.19 16:09:09.012977 [ 546 ] <Fatal> BaseDaemon: 4. raise in ?
[ip-172-31-92-10] 2022.08.19 16:09:09.013300 [ 546 ] <Fatal> BaseDaemon: 5. abort in ?
[ip-172-31-92-10] 2022.08.19 16:09:09.013472 [ 546 ] <Fatal> BaseDaemon: 6. ? in ?
[ip-172-31-92-10] 2022.08.19 16:09:09.013751 [ 546 ] <Fatal> BaseDaemon: 7. ? in ?
[ip-172-31-92-10] 2022.08.19 16:09:09.321887 [ 546 ] <Fatal> BaseDaemon: 8.1. inlined from /build/build_docker/../src/Common/HashTable/HashTableKeyHolder.h:93: keyHolderPersistKey(DB::ArenaKeyHolder&)
[ip-172-31-92-10] 2022.08.19 16:09:09.322250 [ 546 ] <Fatal> BaseDaemon: 8.2. inlined from ../src/Common/HashTable/HashTable.h:950: void HashTable<StringRef, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >::emplaceNonZeroImpl<DB::ArenaKeyHolder&>(unsigned long, DB::ArenaKeyHolder&, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >*&, bool&, unsigned long)
[ip-172-31-92-10] 2022.08.19 16:09:09.322582 [ 546 ] <Fatal> BaseDaemon: 8.3. inlined from ../src/Common/HashTable/HashTable.h:990: void HashTable<StringRef, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >::emplaceNonZero<DB::ArenaKeyHolder&>(DB::ArenaKeyHolder&, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >*&, bool&, unsigned long)
[ip-172-31-92-10] 2022.08.19 16:09:09.322851 [ 546 ] <Fatal> BaseDaemon: 8.4. inlined from ../src/Common/HashTable/HashTable.h:1058: void HashTable<StringRef, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >::emplace<DB::ArenaKeyHolder&>(DB::ArenaKeyHolder&, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >*&, bool&, unsigned long)
[ip-172-31-92-10] 2022.08.19 16:09:09.323192 [ 546 ] <Fatal> BaseDaemon: 8.5. inlined from ../src/Common/HashTable/HashTable.h:1049: void HashTable<StringRef, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >::emplace<DB::ArenaKeyHolder&>(DB::ArenaKeyHolder&, ClearableHashTableCell<StringRef, HashSetCellWithSavedHash<StringRef, DefaultHash<StringRef>, ClearableHashSetState> >*&, bool&)
[ip-172-31-92-10] 2022.08.19 16:09:09.323531 [ 546 ] <Fatal> BaseDaemon: 8.6. inlined from ../src/Common/ColumnsHashingImpl.h:209: DB::ColumnsHashing::columns_hashing_impl::EmplaceResultImpl<void> DB::ColumnsHashing::columns_hashing_impl::HashMethodBase<DB::ColumnsHashing::HashMethodString<StringRef, void, true, false, false>, StringRef, void, false, false>::emplaceImpl<ClearableHashSetWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >, DB::ArenaKeyHolder>(DB::ArenaKeyHolder&, ClearableHashSetWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >&)
[ip-172-31-92-10] 2022.08.19 16:09:09.323884 [ 546 ] <Fatal> BaseDaemon: 8.7. inlined from ../src/Common/ColumnsHashingImpl.h:158: DB::ColumnsHashing::columns_hashing_impl::EmplaceResultImpl<void> DB::ColumnsHashing::columns_hashing_impl::HashMethodBase<DB::ColumnsHashing::HashMethodString<StringRef, void, true, false, false>, StringRef, void, false, false>::emplaceKey<ClearableHashSetWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> > >(ClearableHashSetWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >&, unsigned long, DB::Arena&)
[ip-172-31-92-10] 2022.08.19 16:09:09.324165 [ 546 ] <Fatal> BaseDaemon: 8. ../src/Processors/Transforms/DistinctSortedTransform.cpp:204: bool DB::DistinctSortedTransform::buildFilter<DB::SetMethodString<ClearableHashSetWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> > > >(DB::SetMethodString<ClearableHashSetWithSavedHash<StringRef, DefaultHash<StringRef>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> > >&, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> > const&, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*> > const&, DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul>&, unsigned long, DB::SetVariantsTemplate<DB::ClearableSet>&) const in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:09.538377 [ 546 ] <Fatal> BaseDaemon: 9. /build/build_docker/../src/Processors/Transforms/DistinctSortedTransform.cpp:149: DB::DistinctSortedTransform::transform(DB::Chunk&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:09.694192 [ 546 ] <Fatal> BaseDaemon: 10. /build/build_docker/../src/Processors/ISimpleTransform.h:33: DB::ISimpleTransform::transform(DB::Chunk&, DB::Chunk&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:09.831237 [ 546 ] <Fatal> BaseDaemon: 11. /build/build_docker/../src/Processors/ISimpleTransform.cpp:89: DB::ISimpleTransform::work() in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:09.936182 [ 546 ] <Fatal> BaseDaemon: 12. /build/build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:47: DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:10.036130 [ 546 ] <Fatal> BaseDaemon: 13. /build/build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:92: DB::ExecutionThreadContext::executeTask() in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:10.218051 [ 546 ] <Fatal> BaseDaemon: 14. /build/build_docker/../src/Processors/Executors/PipelineExecutor.cpp:228: DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:10.401008 [ 546 ] <Fatal> BaseDaemon: 15. /build/build_docker/../src/Processors/Executors/PipelineExecutor.cpp:194: DB::PipelineExecutor::executeSingleThread(unsigned long) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:10.572148 [ 546 ] <Fatal> BaseDaemon: 16. /build/build_docker/../src/Processors/Executors/PipelineExecutor.cpp:367: DB::PipelineExecutor::executeImpl(unsigned long) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:10.753624 [ 546 ] <Fatal> BaseDaemon: 17. /build/build_docker/../src/Processors/Executors/PipelineExecutor.cpp:89: DB::PipelineExecutor::execute(unsigned long) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:10.914839 [ 546 ] <Fatal> BaseDaemon: 18. /build/build_docker/../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:79: DB::threadFunction(DB::PullingAsyncPipelineExecutor::Data&, std::__1::shared_ptr<DB::ThreadGroupStatus>, unsigned long) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:11.074600 [ 546 ] <Fatal> BaseDaemon: 19. /build/build_docker/../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:108: DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0::operator()() const in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:11.235371 [ 546 ] <Fatal> BaseDaemon: 20. /build/build_docker/../contrib/libcxx/include/type_traits:3648: decltype(static_cast<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&>(fp)()) std::__1::__invoke_constexpr<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:11.396289 [ 546 ] <Fatal> BaseDaemon: 21. /build/build_docker/../contrib/libcxx/include/tuple:1595: decltype(auto) std::__1::__apply_tuple_impl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:11.557120 [ 546 ] <Fatal> BaseDaemon: 22. /build/build_docker/../contrib/libcxx/include/tuple:1604: decltype(auto) std::__1::apply<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:11.699106 [ 546 ] <Fatal> BaseDaemon: 23. /build/build_docker/../src/Common/ThreadPool.h:187: ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()::operator()() in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:11.859716 [ 546 ] <Fatal> BaseDaemon: 24. /build/build_docker/../contrib/libcxx/include/type_traits:3640: decltype(static_cast<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(fp)()) std::__1::__invoke<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.020001 [ 546 ] <Fatal> BaseDaemon: 25. /build/build_docker/../contrib/libcxx/include/__functional/invoke.h:62: void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&>(ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.179295 [ 546 ] <Fatal> BaseDaemon: 26. /build/build_docker/../contrib/libcxx/include/__functional/function.h:230: std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>::operator()() in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.338719 [ 546 ] <Fatal> BaseDaemon: 27. /build/build_docker/../contrib/libcxx/include/__functional/function.h:711: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.407616 [ 546 ] <Fatal> BaseDaemon: 28. /build/build_docker/../contrib/libcxx/include/__functional/function.h:843: std::__1::__function::__policy_func<void ()>::operator()() const in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.468924 [ 546 ] <Fatal> BaseDaemon: 29. /build/build_docker/../contrib/libcxx/include/__functional/function.h:1184: std::__1::function<void ()>::operator()() const in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.540244 [ 546 ] <Fatal> BaseDaemon: 30. /build/build_docker/../src/Common/ThreadPool.cpp:281: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.630099 [ 546 ] <Fatal> BaseDaemon: 31. /build/build_docker/../src/Common/ThreadPool.cpp:143: void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()::operator()() const in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.723844 [ 546 ] <Fatal> BaseDaemon: 32. /build/build_docker/../contrib/libcxx/include/type_traits:3640: decltype(static_cast<void>(fp)()) std::__1::__invoke<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.818801 [ 546 ] <Fatal> BaseDaemon: 33. /build/build_docker/../contrib/libcxx/include/thread:283: void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>&, std::__1::__tuple_indices<>) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.913081 [ 546 ] <Fatal> BaseDaemon: 34. /build/build_docker/../contrib/libcxx/include/thread:293: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) in /workspace/clickhouse
[ip-172-31-92-10] 2022.08.19 16:09:12.913298 [ 546 ] <Fatal> BaseDaemon: 35. ? in ?
[ip-172-31-92-10] 2022.08.19 16:09:12.913451 [ 546 ] <Fatal> BaseDaemon: 36. __clone in ?
[ip-172-31-92-10] 2022.08.19 16:09:14.227778 [ 546 ] <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read. (calculated checksum: 561828D9BDB179317C1A3A606E634826)
```

**Comment by nikitamikhaylov** [comment]
Simpler query
```
SELECT DISTINCT
    1,
    JSONExtractRaw(toString(rand()), 'x') AS res
FROM numbers(1)
ORDER BY res ASC;
```
The problem is that we use ClearableHashSet for calculating DISTINCT and it doesn't allow to store zeros e.g. empty strings

**Comment by devcrafter** [comment]
Fixed in #44856
